src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+        public int UniqueThreadFlowLocations => _uniqueThreadFlows.Count;
+
+        public int TotalLocations { get; private set; }
+        public int UniqueLocations { get; private set; }
+
+        public int? MessageLengthLimitBytes { get; set; }
+        public bool RemoveCodeFlows { get; set; }
+        public bool RemoveRelatedLocations { get; set; }
+        public bool RemoveGraphs { get; set; }
+        public bool RemoveGraphTraversals { get; set; }
+        public bool RemoveStacks { get; set; }
+        public bool RemoveWebRequests { get; set; }
+        public bool RemoveWebResponses { get; set; }
+
+        public SarifConsolidator(Run run)
+        {
+            _run = run;
+            _uniqueThreadFlows = new Dictionary<ThreadFlowLocation, int>(ThreadFlowLocation.ValueComparer);
+
+            if (_run.Artifacts != null)
+            {
+                // Realize Deferred Artifact list
+                _run.Artifacts = _run.Artifacts?.ToList();
+
+                // Remove indices from Run.Artifact array (redundant)
+                foreach (Artifact a in _run.Artifacts)
+                {
+                    if (a.Location != null)
+                    {
+                        a.Location.Index = -1;
+                    }
+                }
+            }
+
+            // Ensure Run TFL list is ready for consolidating TFLs
+            _run.ThreadFlowLocations = _run.ThreadFlowLocations ?? new List<ThreadFlowLocation>();
+
+            // Add existing TFLs to map (can't consolidate; there may be existing references to them)
+            for (int i = 0; i < _run.ThreadFlowLocations.Count; ++i)
+            {
+                ThreadFlowLocation tfl = _run.ThreadFlowLocations[i];
+                Trim(tfl.Location);
+                _uniqueThreadFlows[tfl] = i;
+            }
+        }
+
+        public void Trim(Result result)
+        {
+            if (result == null) { return; }
+
+            // Remove Result components if configured to
+            result.CodeFlows = (this.RemoveCodeFlows ? null : result.CodeFlows);
+            result.RelatedLocations = (this.RemoveRelatedLocations ? null : result.RelatedLocations);
+            result.Graphs = (this.RemoveGraphs ? null : result.Graphs);
+            result.GraphTraversals = (this.RemoveGraphTraversals ? null : result.GraphTraversals);
+            result.Stacks = (this.RemoveStacks ? null : result.Stacks);
+            result.WebRequest = (this.RemoveWebRequests ? null : result.WebRequest);
+            result.WebResponse = (this.RemoveWebResponses ? null : result.WebResponse);
+
+            // Truncate long Messages if configured
+            if (this.MessageLengthLimitBytes > 0)
+            {
+                if (result?.Message?.Text?.Length > this.MessageLengthLimitBytes)
+                {
+                    result.Message.Text = result.Message.Text.Substring(0, this.MessageLengthLimitBytes.Value);
```

Comment:
>result [](start = 20, length = 6)

We should add some suffix to the string that indicates text was truncated.  'We found a horrible problem and have gone on and on about it. Here's more detail you likely didn't [remainder of message truncated due to exceeding size limit...].

Reply:
Added ellipse along with Markdown length check per Larry.